sdk/core/Azure.Core/src/HttpMessage.cs

```diff
@@ -13,93 +13,86 @@ namespace Azure.Core
     /// <summary>
     /// Represents a context flowing through the <see cref="HttpPipeline"/>.
     /// </summary>
-    public sealed class HttpMessage : IDisposable
+    public sealed class HttpMessage : PipelineMessage
     {
-        private ArrayBackedPropertyBag<ulong, object> _propertyBag;
-        private Response? _response;
-
         /// <summary>
         /// Creates a new instance of <see cref="HttpMessage"/>.
         /// </summary>
         /// <param name="request">The request.</param>
         /// <param name="responseClassifier">The response classifier.</param>
         public HttpMessage(Request request, ResponseClassifier responseClassifier)
+            : base(request)
         {
-            Argument.AssertNotNull(request, nameof(Request));
-            Request = request;
+            Argument.AssertNotNull(request, nameof(request));
+
             ResponseClassifier = responseClassifier;
-            BufferResponse = true;
-            _propertyBag = new ArrayBackedPropertyBag<ulong, object>();
         }
 
         /// <summary>
         /// Gets the <see cref="Request"/> associated with this message.
         /// </summary>
-        public Request Request { get; }
+        public new Request Request { get => (Request)base.Request; }
 
         /// <summary>
         /// Gets the <see cref="Response"/> associated with this message. Throws an exception if it wasn't set yet.
         /// To avoid the exception use <see cref="HasResponse"/> property to check.
         /// </summary>
-        public Response Response
+        public new Response Response
         {
             get
             {
-                if (_response == null)
+                if (base.Response is null)
                 {
 #pragma warning disable CA1065 // Do not raise exceptions in unexpected locations
-                    throw new InvalidOperationException("Response was not set, make sure SendAsync was called");
+                    throw new InvalidOperationException($"{nameof(Response)} is not set on this message. "
+                        + "This is may be because the message was not sent via pipeline.Send, "
+                        + "the pipeline transport did not populate the response, or because "
+                        + $"{nameof(ExtractResponse)} was called. You can check the {nameof(HasResponse)} "
+                        + "property to test whether the message has a response value before "
+                        + $"accessing the {nameof(Response)} property.");
 #pragma warning restore CA1065 // Do not raise exceptions in unexpected locations
                 }
-                return _response;
+                return (Response)base.Response;
             }
-            set => _response = value;
+
+            set => base.Response = value;
```

Comment:
We don't need an adapter here to adapt from PipelineResponse to Response because on the ClientModel type, the Response setter is protected, so it can only be set from subtypes.  This means the only way to set it for this type is through this setter, which takes Response.